raiden/network/transport/matrix/rtc/web_rtc.py

```diff
@@ -488,188 +510,179 @@ def _initialize_web_rtc(self, partner_address: Address) -> None:
-
-            self._address_to_rtc_partners[partner_address] = _RTCPartner(
-                partner_address,
-                self.node_address,
-                self._handle_message,
-                self._handle_candidates_callback,
-                self._handle_closed_connection,
-            )
-
-        return self._address_to_rtc_partners[partner_address]
+    def _add_connection(self, partner_address: Address, conn: _RTCConnection) -> None:
+        assert (
+            len(self._address_to_connections[partner_address]) < 2
+        ), "cannot have more than 2 connections per partner"
+        assert (
+            conn.call_id not in self._address_to_connections[partner_address]
+        ), "must not be there already"
+        self._address_to_connections[partner_address][conn.call_id] = conn
+
+    def _get_connection(self, partner_address: Address, call_id: str = None) -> _RTCConnection:
+        conns = self._address_to_connections[partner_address]
+        if call_id is not None:
+            return conns[call_id]
+        return next(iter(conns.values()))
 
     def has_ready_channel(self, partner_address: Address) -> bool:
-        if partner_address not in self._address_to_rtc_partners:
+        conns = self._address_to_connections.get(partner_address)
+        if conns is None:
             return False
-        channel = self._address_to_rtc_partners[partner_address].channel
-        if channel is None:
-            return False
-        if channel.readyState == RTCChannelState.OPEN.value:
-            return True
-        return False
+        return any(
+            conn.channel is not None and conn.channel.readyState == _RTCChannelState.OPEN.value
+            for conn in conns.values()
+        )
 
     def _reset_state(self) -> None:
-        self._address_to_rtc_partners = {}
+        self._address_to_connections = {}
+
+    def _have_both_connections(self, partner_address: Address) -> bool:
+        return (
+            partner_address in self._address_to_connections
+            and len(self._address_to_connections[partner_address]) == 2
+        )
 
     def _set_candidates_for_address(
         self, partner_address: Address, content: Dict[str, Any]
     ) -> None:
-        rtc_partner = self.get_rtc_partner(partner_address)
-        self.schedule_task(rtc_partner.set_candidates(content))
+        if not self._have_both_connections(partner_address):
+            conn = self._get_connection(partner_address, content["call_id"])
+            self.schedule_task(conn.set_candidates(content))
 
     def _process_signalling_for_address(
-        self, partner_address: Address, description: Dict[str, str]
+        self, partner_address: Address, rtc_message_type: str, description: Dict[str, str]
     ) -> None:
-        rtc_partner = self.get_rtc_partner(partner_address)
+        if rtc_message_type == _RTCMessageType.OFFER.value:
+            if self._have_both_connections(partner_address):
```

Comment:
Couldn't we prevent creating a second channel if there is already one available?

Reply:
Sure, but then we're back to only allowing one channel per peer. Which, if it goes down for any reason, means we would need to establish a new channel again, rather than using the other, presumably still available, channel.

Reply:
I think this is not exactly true. If there is a channel alive, while an offer comes in, we can gracefully decline or even ignore the message. This does not imply that only one channel is allowed. Since the process of establishing a channel takes a couple of seconds, two peers could try to create one while the other is not established (channel.readyState == open) yet. In this case we should allow the creation of the two. But if there is a channel available, which means that the other already sees the channel as open (open connection) we can decline the offer. (the offer could come with a bit of latency due to matrix)

Additionally, even if we have two channels and one goes down and we keep using the other one, we end up with having only one channel. Then we are back at the question waht happens if this channel goes down? For this case we should have a retry mechanism in place anyway, so there is no reason to always aim for two channels because if one dies the other eventually will too.

Reply:
Maybe it should be noticed that most of the times there is only one user creating a channel. So it is not the default that there are two channels alive.

Reply:
Ok, changed the code to ignore offers if we already have a channel.